src/lib.rs

```diff
@@ -399,6 +399,38 @@ impl Client {
         self.send(&TimeMetric::new(stat.into().as_ref(), &start_time, &end_time), tags)
     }
 
+    /// Time how long it takes for an async block of code to execute.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    ///   use dogstatsd::{Client, Options};
+    ///   use std::thread;
+    ///   use std::time::Duration;
+    ///
+    ///   async fn timer() {
+    ///       let client = Client::new(Options::default()).unwrap();
+    ///       client.time("timer", &["tag:time"], || async {
+    ///           thread::sleep(Duration::from_millis(200))
+    ///       })
+    ///       .await
+    ///       .unwrap_or_else(|e| println!("Encountered error: {}", e))
+    ///   }
+    /// ```
+    pub async fn async_time<'a, Fn, Fut, Out, I, S, T>(&self, stat: S, tags: I, block: Fn) -> Result<Out, DogstatsdError>
+        where Fn: FnOnce() -> Fut,
+              Fut: Future<Output=Out>,
+              I: IntoIterator<Item=T>,
+              S: Into<Cow<'a, str>>,
+              T: AsRef<str>,
+    {
+        let start_time = Utc::now();
+        let result: Out = block().await;
+        let end_time = Utc::now();
+        self.send(&TimeMetric::new(stat.into().as_ref(), &start_time, &end_time), tags)?;
+        Ok(result)
+    }
+
```

Comment:
You can simplify this by taking a future directly
```suggestion
    /// Time how long it takes for an async block of code to execute.
    ///
    /// # Examples
    ///
    /// ```
    ///   use dogstatsd::{Client, Options};
    ///   use std::thread;
    ///   use std::time::Duration;
    ///
    ///   async fn timer() {
    ///       let client = Client::new(Options::default()).unwrap();
    ///       client.async_time("timer", &["tag:time"], async {
    ///           thread::sleep(Duration::from_millis(200))
    ///       })
    ///       .await
    ///       .unwrap_or_else(|e| println!("Encountered error: {}", e))
    ///   }
    /// ```
    pub async fn async_time<'a, Fut, Out, I, S, T>(&self, stat: S, tags: I, block: Fut) -> Result<Out, DogstatsdError>
        where Fut: Future<Output=Out>,
              I: IntoIterator<Item=T>,
              S: Into<Cow<'a, str>>,
              T: AsRef<str>,
    {
        let start_time = Utc::now();
        let result: Out = block.await;
        let end_time = Utc::now();
        self.send(&TimeMetric::new(stat.into().as_ref(), &start_time, &end_time), tags)?;
        Ok(result)
    }
```